dvc/objects/stage.py

```diff
@@ -1,91 +1,131 @@
 import errno
 import os
 from concurrent.futures import ThreadPoolExecutor
+from functools import partial
 
 from dvc.exceptions import DvcIgnoreInCollectedDirError
 from dvc.hash_info import HashInfo
 from dvc.ignore import DvcIgnore
+from dvc.objects.file import HashFile
 from dvc.progress import Tqdm
 from dvc.utils import file_md5
 
 
-def _get_file_hash(path_info, fs, name):
+def _upload_file(path_info, fs, odb):
+    from dvc.utils import tmp_fname
+    from dvc.utils.stream import HashedStreamReader
+
+    tmp_info = odb.fs.path_info / tmp_fname()
+    with fs.open(path_info, mode="rb", chunk_size=fs.CHUNK_SIZE) as stream:
+        stream = HashedStreamReader(stream)
+        odb.fs.upload_fobj(
+            stream, tmp_info, desc=path_info.name, total=fs.getsize(path_info)
+        )
+
+    obj = HashFile(tmp_info, odb.fs, stream.hash_info)
+    return path_info, obj
+
+
+def _get_file_hash(path_info, fs, name, odb, upload):
+    if upload:
+        assert odb and name == "md5"
+        return _upload_file(path_info, fs, odb)
+
     info = fs.info(path_info)
     if name in info:
         assert not info[name].endswith(".dir")
-        return HashInfo(name, info[name], size=info["size"])
-
-    func = getattr(fs, name, None)
-    if func:
-        return func(path_info)
-
-    if name == "md5":
-        return HashInfo(
+        hash_info = HashInfo(name, info[name], size=info["size"])
+    elif hasattr(fs, name):
+        hash_info = getattr(fs, name)(path_info)
+    elif name == "md5":
+        hash_info = HashInfo(
             name, file_md5(path_info, fs), size=fs.getsize(path_info)
         )
+    else:
+        raise NotImplementedError
 
-    raise NotImplementedError
+    return path_info, HashFile(path_info, fs, hash_info)
 
 
-def get_file_hash(path_info, fs, name, state=None):
+def _get_file_obj(
+    path_info, fs, name, odb=None, state=None, upload=False, **kwargs
+):
     if state:
         hash_info = state.get(  # pylint: disable=assignment-from-none
             path_info, fs,
         )
         if hash_info:
-            return hash_info
+            return path_info, HashFile(path_info, fs, hash_info)
 
     if not fs.exists(path_info):
         raise FileNotFoundError(
             errno.ENOENT, os.strerror(errno.ENOENT), path_info
         )
 
-    hash_info = _get_file_hash(path_info, fs, name)
+    target_info, obj = _get_file_hash(
+        path_info, fs, name, odb=odb, upload=upload
+    )
 
+    hash_info = obj.hash_info
     if state:
         assert ".dir" not in hash_info.value
-        state.save(path_info, fs, hash_info)
+        state.save(obj.path_info, obj.fs, obj.hash_info)
```

Comment:
This one should actuallly stay the same, as this is us saving information about the original path_info and fs, notabout the temporary file on odb.fs